src/python/pants/engine/internals/scheduler.py

```diff
@@ -409,61 +384,74 @@ def execution_request(
         invalidated = self._scheduler.invalidate_all_files()
         self._maybe_visualize()
         return invalidated
 
-    def node_count(self):
-        return self._scheduler.graph_len()
-
-    def metrics(self) -> Dict[str, int]:
+    def metrics(self) -> dict[str, int]:
         """Returns metrics for this SchedulerSession as a dict of metric name to metric value."""
-        return cast(Dict[str, int], self._scheduler._metrics(self._session))
+        return native_engine.scheduler_metrics(self.py_scheduler, self.py_session)
 
-    def _maybe_visualize(self):
+    def _maybe_visualize(self) -> None:
         if self._scheduler.visualize_to_dir is not None:
             name = f"graph.{self._run_count:03d}.dot"
             self._run_count += 1
             self.visualize_graph_to_file(os.path.join(self._scheduler.visualize_to_dir, name))
 
     def teardown_dynamic_ui(self) -> None:
-        native_engine.teardown_dynamic_ui(self._scheduler._scheduler, self._session)
+        native_engine.teardown_dynamic_ui(self.py_scheduler, self.py_session)
 
-    def execute(self, execution_request: ExecutionRequest):
+    def execute(
+        self, execution_request: ExecutionRequest
+    ) -> tuple[tuple[tuple[Any, Return], ...], tuple[tuple[Any, Throw], ...]]:
         """Invoke the engine for the given ExecutionRequest, returning Return and Throw states.
 
         :return: A tuple of (root, Return) tuples and (root, Throw) tuples.
         """
         start_time = time.time()
-        roots = list(
-            zip(
-                execution_request.roots,
-                self._scheduler._run_and_return_roots(self._session, execution_request.native),
-            ),
-        )
+        try:
+            raw_roots = native_engine.scheduler_execute(
+                self.py_scheduler,
+                self.py_session,
+                execution_request.native,
+            )
+        except native_engine.PollTimeout:
+            raise ExecutionTimeoutError("Timed out")
 
-        self._maybe_visualize()
+        states = [
+            Throw(
+                raw_root.result(),
+                python_traceback=raw_root.python_traceback(),
+                engine_traceback=raw_root.engine_traceback(),
+            )
+            if raw_root.is_throw()
+            else Return(raw_root.result())
+            for raw_root in raw_roots
+        ]
 
+        roots = list(zip(execution_request.roots, states))
+
+        self._maybe_visualize()
         logger.debug(
             "computed %s nodes in %f seconds. there are %s total nodes.",
             len(roots),
             time.time() - start_time,
             self._scheduler.graph_len(),
         )
 
-        returns = tuple((root, state) for root, state in roots if type(state) is Return)
-        throws = tuple((root, state) for root, state in roots if type(state) is Throw)
-        return cast(Tuple[Tuple[Return, ...], Tuple[Throw, ...]], (returns, throws))
```

Comment:
This return type was wrong. The method signature is now correct, and MyPy was able to figure it out :)